cmake/CMakeLists.txt

```diff
@@ -779,28 +803,30 @@ if (onnxruntime_BUILD_BENCHMARKS)
   if (TARGET benchmark::benchmark)
     message("Use benchmark from preinstalled system lib")
   else()
-    message("Use benchmark from submodule")
     # We will not need to test benchmark lib itself.
     set(BENCHMARK_ENABLE_TESTING OFF CACHE BOOL "Disable benchmark testing as we don't need it.")
     # We will not need to install benchmark since we link it statically.
     set(BENCHMARK_ENABLE_INSTALL OFF CACHE BOOL "Disable benchmark install to avoid overwriting vendor install.")
-    add_subdirectory(${PROJECT_SOURCE_DIR}/external/googlebenchmark EXCLUDE_FROM_ALL)
+   
+      FetchContent_Declare(
+        google_benchmark
+        URL ${DEP_URL_google_benchmark}
+      )
+      FetchContent_MakeAvailable(google_benchmark)
+   
     #add_library(benchmark::benchmark ALIAS benchmark)
   endif()
 endif()
 
-if (NOT WIN32)
-  if (onnxruntime_PREFER_SYSTEM_LIB)
-    find_package(nsync)
-  endif()
-  if (TARGET nsync_cpp)  # linking error with nsync_FOUND (why?)
-    message("Use nsync from preinstalled system lib")
-  else()
-    message("Use nsync from submodule")
-    add_subdirectory(${PROJECT_SOURCE_DIR}/external/nsync EXCLUDE_FROM_ALL)
-  endif()
+if (NOT WIN32)           
+    FetchContent_Declare(
+    google_nsync
+    URL ${DEP_URL_google_nsync}    
+    FIND_PACKAGE_ARGS NAMES nsync
+    )
+    FetchContent_MakeAvailable(google_nsync)
+	set(nsync_SOURCE_DIR ${google_nsync_SOURCE_DIR})   
 endif()
-# External dependencies
 list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/external)
 
 #protobuf begin
```

Comment:
If they also use cmake FetchContent or FindPackage module, it will be good. On the other hand, if they use git submodules and "add_subdirectory" like what we have now, it won't work.

Reply:
Actually, FetchContent mechanism is now under the hood of find_package as well. More integrated.

What I meant by this comment is that XNNPACK has this check:

[https://github.com/google/XNNPACK/blob/003c580e696a774afdc984996ee909b7c8d8128c/CMakeLists.txt#L124](https://github.com/google/XNNPACK/blob/003c580e696a774afdc984996ee909b7c8d8128c/CMakeLists.txt#L124)

So if we set the var, then it would re-use ours. It is done on purpose. The same is true for some other components.